lib/queuePopulator/IngestionPopulator.js

```diff
@@ -118,47 +116,46 @@ class IngestionPopulator {
     }
 
     /**
-     * Close provisioned ingestion source
-     * @param {function} cb - callback function
+     * add an Ingestion Reader
+     * @param {Object} source - new source object
+     * @param {Function} cb - callback(error)
      * @return {undefined}
      */
-    close(cb) {
-        return this._closeLogState(cb);
-    }
-
-    _setupLogSources() {
-        this.ingestionConfig.sources.map(sourceObj => {
+    addNewLogSource(source, cb) {
+        this._extension.createZkPath(err => {
+            if (err) {
+                return cb(err);
+            }
             const zookeeperUrlIngest =
                 `${this.zkConfig.connectionString}` +
-                `${this.ingestionConfig.zookeeperPath}/` +
-                `${sourceObj.name}`;
-            const zkEndpointIngest =
-                `${zookeeperUrlIngest}/raft-id-dispatcher`;
-            this._subscribeToLogSourceDispatcher(zkEndpointIngest,
-                sourceObj, this.ingestionConfig.zookeeperPath);
-            return undefined;
-        });
+                `${this.ingestionConfig.zookeeperPath}/${source.name}`;
+            const zkEndpointIngest = `${zookeeperUrlIngest}/raft-id-dispatcher`;
+            return this._subscribeToLogSourceDispatcher(zkEndpointIngest,
+                source, cb);
+        }, source);
     }
 
-    _subscribeToLogSourceDispatcher(zkEndpoint, bucketd, ingestionPath) {
-        // NOTE: Bug where raftIdDispatcher will just be overwritten if more
-        // than one source exists. This will change in a later commit though
-        // when supporting workflow manager
-        this.raftIdDispatcher =
+    _subscribeToLogSourceDispatcher(zkEndpoint, bucketd, cb) {
+        const ingestionPath = this.ingestionConfig.zookeeperPath;
+        // TODO: Replace ProvisionDispatcher?
+        const raftIdDispatcher =
             new ProvisionDispatcher({ connectionString: zkEndpoint });
-        this.raftIdDispatcher.subscribe((err, items) => {
+        raftIdDispatcher.subscribe((err, items) => {
             if (err) {
                 this.log.error(
                     'error when receiving log source provision list',
                     { zkEndpoint, error: err });
-                return undefined;
+                return cb(err);
             }
+            const key = bucketd.name;
+            this._provisionedIngestionSources[key] = raftIdDispatcher;
             if (items.length === 0) {
                 this.log.info('no log source provisioned, idling',
                               { zkEndpoint });
             }
             const newRaftReaders = items.map(
                 token => {
+                    // TODO: will token ever be mongo?
```

Comment:
I think Lifecycle moved away from using a dispatcher, so should we since there is only one process handling the ingestion from source. This token part can be removed as we will always ingest from a RaftLog.

Reply:
awesome! removing this part then